### ISSUE TITLE
Feature request / question: copyright year linting

### ISSUE THREAD

**Issue by jayaddison** [sphinx-contrib/sphinx-lint#53]
This may not be the correct place for a feature like this, so please let me know - and feel free to close this issue - if not.

I'm wondering whether checking the configured [Sphinx copyright year](https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-copyright) would be something that `sphinx-lint` could/should include.

I'm willing to attempt an implementation if so, and would be glad for any recommendations on alternative approaches or projects that could include this (I've looked for suitable extensions, but not found any so far) if not.

**Comment by hugovk** [comment]
What determines a correct/incorrect copyright line?

If it's about the year, many projects don't include the current year, and some also skip the initial year. 

For example, Google, Microsoft, Facebook, Netflix, Linux Foundation, Amazon:

https://hynek.me/til/copyright-years/

Conversely, CPython always includes the current year:

```python
copyright = '2001-%s, Python Software Foundation' % time.strftime('%Y')
```

https://github.com/python/cpython/blob/2f2fa03ff3d566b675020787e23de8fb4ca78e99/Doc/conf.py#L52

**Comment by jayaddison** [comment]
> What determines a correct/incorrect copyright line?

Yep, good question.  Perhaps the best answer is... a legal discussion determines correctness, if it becomes an important factor in some other proceedings?  So a linter could only be a hint to help developers (particularly the ones in environments where including-current-year is recommended).

Based on that -- environment-tuned guidance, not correctness -- a linter rule could be progressively configurable, with sensible defaults.  "Contains current year" as a default might introduce some false positives, but could be tuned down to "Contains a year >= <start-year>", or similar, by configuration.

**Comment by hugovk** [comment]
Linters tend to influence behaviour, so since all options (start and end years, start year only, no years) are correct, perhaps it should be opt-in to enforce a given flavour?

**Comment by jayaddison** [comment]
That seems good, yep - that'd also minimize the number of linter warnings introduced for most projects if-and-when such a rule became available.

One suggestion/exception to that, though: perhaps a copyright-configuration-exists check (that tests for a non-empty `copyright` configuration entry) would make sense?

**Comment by JulienPalard** [comment]
> perhaps a copyright-configuration-exists check (that tests for a non-empty copyright configuration entry) would make sense?

I'm not even sure, but IANAL. A documentation without a copyright would look OK to me, as it would be protected by the [Berne Convention](https://en.wikipedia.org/wiki/Berne_Convention).

What I fear is that the answer to this thread is "it depends on which country the writer is", in which case better not implement the test...

**Comment by jayaddison** [comment]
> What I fear is that the answer to this thread is "it depends on which country the writer is"

That matches my understanding too, yep.

Until a better/clearer/universally-applicable answer arrives, then it'd seem safe to leave any such linting rules opt-in.

To perhaps explain a bit more context here:

My motivating factor for opening this feature request was to try to figure out how a site that uses (or would like to use) continuous integration to lint and publish Sphinx documentation could include a year range in their copyright configuration while also ensuring that the build is reproducible.

To me, a good way to achieve that would be to re-use linting (a an existing and widely-used inspection mechanism) to detect out-of-date copyright (with, I admit, very little specification of what those checks might involve).

Perhaps an equally reasonable answer would be something along the lines of "add a bash script that uses grep" -- I think that's fine, although I also like bundling relevant functionality together, especially where it could improve portability.

**Comment by jayaddison** [comment]
(I probably shouldn't even be the person requesting the feature, to be honest!  it seemed like a good idea to me, but increasingly I wonder whether I'm just roaming the internet trying to find obscure problems that no-one else seems to be addressing (perhaps for various entirely reasonable factors).  the goal is certainly to improve quality and save time overall, but sometimes it's tricky to figure out whether the activation energy required to jump a hurdle is worth it -- or whether it's a good idea to be on the other side of the hurdle in the first place)

**Comment by jayaddison** [comment]
Sorta-related: sphinx-doc/sphinx#3463 (`sphinx` itself validates the presence of copyright configuration for EPUB3 files)

**Comment by jayaddison** [comment]
I don't know whether I'll implement this, or whether it's worth implementing.  But I've been thinking about it, and would like to document some of that.

#### Draft of a possible design
- Add a checker for copyright-year, disabled by default, with configurable `expectation` options for...
  - `current_year` (boolean, default false)
  - `start_year` (optional, int, default empty)
  - `entity_name` (optional, string, default empty)
- Filter the checker to files named `conf.py` during file walking
- Allow the checker to self-configure from a `tool.sphinx-lint` section in the `conf.py` file(s) (somewhat analogous to the way that some Python linters can read their configuration from `pyproject.toml`)
- Validate the contents of the `copyright` field based on the options - ignoring any dynamic Python code

#### Oddities
It's slightly unusual that the checker would be reading configuration from the same file that it is linting.

#### Longer-term: configuration discovery
Linter configuration could be applicable to other checkers in future -- so it _might_ make sense to discover all `conf.py` files (frequently there will only be one) in a first-pass tree-walk, before then applying each retrieved configuration to adjacent-and-deeper files during the lint-check tree-walk.

Whether that would make sense to implement during an initial implementation seems debatable.

**Comment by jayaddison** [comment]
> Allow the checker to self-configure from a tool.sphinx-lint section in the `conf.py` file(s) (somewhat analogous to the way that some Python linters can read their configuration from `pyproject.toml`)

This doesn't make much sense; I wasn't thinking straight when I wrote that.  Since the file is Python code, it doesn't really have config sections like a TOML file / INI file / etc.

As a general rule, I don't think that linters should evaluate the code that they're checking.  So that implies either reading config from some specially-marked-up comments in the config, or reading the settings from elsewhere.

**Comment by jayaddison** [comment]
Ok, sorta-tempted though I am to try an implementation, I'm going to attempt something that I don't usually do: choosing _not_ to do something.  No-one seems to be asking for this feature, and the implementation would be a bit different to the other checks in `sphinx-lint` (that are generally markup-focused, not configuration-focused).

Always possible that someone will take another look at it in future, but let's save some effort and not do this at the moment.
